File: backend/src/service/agent-tools/article.tools.ts
import { tool } from "langchain";
// 👆 引入一个“工具模具”，用它造出来的工具，AI才认得
import { z } from "zod";
// 👆 引入“安检门”，用来校验AI传进来的参数格式对不对
import { searchArticles, getArticleTitles } from '@/service/search.service';
// 👆 引入你原本写好的业务函数（真正查数据库干活的）

// 工具1
export const searchArticlesTool = tool(
    async ({ keyword, page, size }) => {
        const result = await searchArticles({ keyword, page, size });
        const cleanData = result.articlesWithMeta.map(item => ({
            title: item.title,
            content: item.content,
        }));
        return JSON.stringify({
            total: result.total,
            articles: cleanData,
            note: "请根据上述文章列表，用自然语言向用户推荐，不要直接显示 JSON。"
        });
    },
    {
        name: "search_articles",
        description: "根据关键词搜索文章列表...",
        schema: z.object({
            keyword: z.string().describe("搜索关键词"),
            page: z.number().default(1),
            size: z.number().default(10),
        }
        ),
    });

// 工具2：获取文章标题建议
export const getArticleTitlesTool = tool(
    async ({ keyword }) => {
        const titles = await getArticleTitles({ keyword });
        return JSON.stringify({
            titles: titles,
            note: "请根据上述标题列表，用自然语言向用户推荐相关文章标题，不要直接显示 JSON。"
        });
    },
    {
        name: "get_article_titles",
        description: "根据关键词获取相关文章标题建议，用于搜索自动补全",
        schema: z.object({
            keyword: z.string().describe("搜索关键词前缀"),
        }),
    }
);

// 导出文章相关工具数组
export const articleTools = [searchArticlesTool, getArticleTitlesTool];

File: backend/ai-service/tools/article_tools.py
import json

import httpx
from langchain_core.tools import tool

import config
from tools.java_api import extract_data, extract_list

JAVA_URL = config.JAVA_BACKEND_URL


@tool
async def search_articles(keyword: str, page: int = 1, size: int = 10) -> str:
    """根据关键词搜索文章列表，返回匹配的文章标题和内容摘要。当用户想查找某个主题/关键词的文章时使用此工具。"""
    async with httpx.AsyncClient(timeout=config.HTTP_TIMEOUT) as client:
        resp = await client.get(
            f"{JAVA_URL}/api/search",
            params={"keyword": keyword, "page": page, "size": size},
        )
        payload = resp.json()
        # /api/search 的 data 为分页对象 {total, size, page, list: [...]}
        clean_data = [
            {"title": a.get("title"), "content": a.get("content")}
            for a in extract_list(payload)
            if isinstance(a, dict)
        ]
        return json.dumps({
            "total": extract_data(payload).get("total", 0),
            "articles": clean_data,
        }, ensure_ascii=False)


@tool
async def get_article_titles(keyword: str) -> str:
    """根据关键词搜索文章标题列表。当用户问"有没有标题包含XX的文章"、"找标题里带某词的文章"时使用此工具。"""
    async with httpx.AsyncClient(timeout=config.HTTP_TIMEOUT) as client:
        resp = await client.get(
            f"{JAVA_URL}/api/search/titles",
            params={"keyword": keyword},
        )
        # /api/search/titles 的 data 直接是标题字符串数组
        return json.dumps({
            "titles": extract_list(resp.json()),
        }, ensure_ascii=False)


article_tools = [search_articles, get_article_titles]
